iree/compiler/InputConversion/MHLO/MHLOToLinalgOnTensors.cpp

```diff
@@ -268,6 +268,101 @@ struct ConstOpConversion : public OpConversionPattern<mhlo::ConstOp> {
   }
 };
 
+//===----------------------------------------------------------------------===//
+// mhlo.RngUniformOp conversion patterns.
+//===----------------------------------------------------------------------===//
+
+namespace {
+/// Returns an ArrayAttr that contains `nLoops` attributes. All the attributes
+/// are "parallel" except the last `nReduction` elements, where are "reduction"
+/// attributes.
+SmallVector<StringRef, 3> GetParallelAndReductionIterators(
+    unsigned nLoops, unsigned nReduction) {
+  SmallVector<StringRef, 3> res(nLoops - nReduction,
+                                getParallelIteratorTypeName());
+  res.append(nReduction, getReductionIteratorTypeName());
+  return res;
+}
+
+struct RngUniformConversion : public OpConversionPattern<mhlo::RngUniformOp> {
+  using OpConversionPattern<mhlo::RngUniformOp>::OpConversionPattern;
+
+  LogicalResult matchAndRewrite(
+      mhlo::RngUniformOp op, ArrayRef<Value> args,
+      ConversionPatternRewriter &rewriter) const final {
+    Type int32Type = IntegerType::get(op.getContext(), /*width=*/32);
+    auto loc = op.getLoc();
+    auto targetTy = op->getResult(0).getType().dyn_cast<ShapedType>();
+    auto targetRank = targetTy.getRank();
+    Value output;
+    Value initTensor = rewriter.create<linalg::InitTensorOp>(
+        loc, ArrayRef<Value>({}), targetTy.getShape(),
+        targetTy.getElementType());
+    auto resultTy = this->typeConverter->convertType(op.getResult().getType())
+                        .cast<ShapedType>();
+    // Creates index map using target matrix's rank.
+    SmallVector<AffineMap, 3> indexingMaps;
+    indexingMaps.push_back(AffineMap::get(targetRank, /*symbolCount=*/0,
+                                          SmallVector<AffineExpr>({}),
+                                          rewriter.getContext()));
+    indexingMaps.push_back(AffineMap::get(targetRank, /*symbolCount=*/0,
+                                          SmallVector<AffineExpr>({}),
+                                          rewriter.getContext()));
+    SmallVector<AffineExpr> outputExprs;
+    for (int i = 0; i < targetRank; i++) {
+      outputExprs.push_back(rewriter.getAffineDimExpr(i));
+    }
+    indexingMaps.push_back(AffineMap::get(targetRank, /*symbolCount=*/0,
+                                          outputExprs, rewriter.getContext()));
+    // Generic region with LCG Algorithm that make use of element index from:
```

Comment:
Is there a reason to either not use what is already there upstream, or adding this to upstream? From the patch you linked, it seems like that is already a named op. If we either use it or add a new named op, we can just lower to that operation here?

Reply:
This pass generalize to multiple dimensions in the case we need 3d tensors random for example.

Reply:
But this could be added to `linalg` directly right?

Reply:
Hmm since the main purpose for this pass is mostly a mhlo lowering, maybe TF would be more ideal, but not sure if the TF folks will take any uniform the LCG algo if it doesn't stay completely true to there side. But according to Stella, the uniform algo on TF side is also abstract and backend dependent. Hence, at the end I am not quite sure where the best spot would be. 😅

Reply:
Is it common to see N-D tensors random op where `N > 3`? If not, I would add the named op and lower it to named ops, like what we've done for conv ops.

Reply:
for our bert small train mlir, we have mostly mostly 3ds and one 2d, and one 4d.